urnai/sc2/environments/sc2environment.py

```diff
@@ -0,0 +1,123 @@
+import sys
+
+from absl import flags
+from pysc2.env import sc2_env
+from pysc2.env.environment import TimeStep
+from pysc2.lib import actions, features
+
+from urnai.environments.environment_base import EnvironmentBase
+
+
+class SC2Env(EnvironmentBase):
+    def __init__(
+            self,
+            map_name='Simple64',
+            player_race=sc2_env.Race.random,
+            enemy_race=sc2_env.Race.random,
+            difficulty=sc2_env.Difficulty.very_easy,
+            visualize=False,
+            reset_done=True,
+            spatial_dim=16,
+            step_mul=16,
+            game_steps_per_episode=0,
+            realtime=False,
+            self_play=False,
+            action_space=actions.ActionSpace.RAW,
+            use_raw_units=True,
+            raw_resolution=64,
+            use_feature_units=True,
+            screen=64,
+            minimap=64,
+    ):
+        super().__init__(map_name, visualize, reset_done)
+
+        FLAGS = flags.FLAGS
+        FLAGS(sys.argv)
+
+        self.self_play = self_play
+        self.step_mul = step_mul
+        self.game_steps_per_episode = game_steps_per_episode
+        self.spatial_dim = spatial_dim
+        self.player_race = player_race
+        self.enemy_race = enemy_race
+        self.difficulty = difficulty
+        if self.self_play:
+            self.players = [
+                sc2_env.Agent(self.player_race),
+                sc2_env.Agent(self.enemy_race),
+            ]
+        else:
+            self.players = [
+                sc2_env.Agent(self.player_race),
+                sc2_env.Bot(self.enemy_race, self.difficulty),
+            ]
```

Comment:
What if I want to use more than two races?

Reply:
Do you mean more than two races in a single game? Like Zergs vs Terran vs Protoss (3 players or more against each other)?

Reply:
Both.

Reply:
I searched a bit into the pysc2 code implementation of their environment, and apparently they only have support to at most 2 players on a single map. Here, `players` is the list of Agents or Bots passed as a parameter to their initialization method:

```python
for p in players:
  if not isinstance(p, (Agent, Bot)):
    raise ValueError(
        "Expected players to be of type Agent or Bot. Got: %s." % p)

num_players = len(players)
self._num_agents = sum(1 for p in players if isinstance(p, Agent))
self._players = players

if not 1 <= num_players <= 2 or not self._num_agents:
  raise ValueError("Only 1 or 2 players with at least one agent is "
                    "supported at the moment.")
```

Reply:
But even with at most 2 players at a time, our current code supports the following:
- One Agent against one Bot
- One Agent against another Agent

For the second case, should our environment return two instances of observations and rewards on the method `step()`, for example? Because I believe our current implementation only returns the observation and reward of the first Agent.

Reply:
What a sad thing to allow only 2 agents. I don't know how pysc2 will return the observations in this case, I believe the agents should have an identifier to differentiate them. For now, let's just focus on one Agent and later we can implement two Agents. @RickFqt create a ticket about this, please.

Reply:
> @RickFqt create a ticket about this, please.

Do you mean creating a new issue?

Reply:
Yes.